diff --git a/toolchains/esp32/sdkconfig b/toolchains/esp32/sdkconfig
--- a/toolchains/esp32/sdkconfig
+++ b/toolchains/esp32/sdkconfig
@@ -1206,7 +1206,7 @@ CONFIG_ESP32_ECO3_CACHE_LOCK_FIX=y
 #
 # Non-backward compatible options
 #
-# CONFIG_ESP_SYSTEM_ESP32_SRAM1_REGION_AS_IRAM is not set
+CONFIG_ESP_SYSTEM_ESP32_SRAM1_REGION_AS_IRAM=y
 # end of Non-backward compatible options
 # end of Memory
 
