diff --git a/src/compiler/third_party/semver/semver.c b/src/compiler/third_party/semver/semver.c
index 29bc1868d..908068579 100644
--- a/src/compiler/third_party/semver/semver.c
+++ b/src/compiler/third_party/semver/semver.c
@@ -152,9 +152,6 @@ semver_parse (const char *str, semver_t *ver) {
 
   res = semver_parse_version(buf, ver);
   free(buf);
-#if DEBUG > 0
-  printf("[debug] semver.c %s = %d.%d.%d, %s %s\n", str, ver->major, ver->minor, ver->patch, ver->prerelease, ver->metadata);
-#endif
   return res;
 }
 
